src/extractor/extractor.cpp

```diff
@@ -252,58 +254,60 @@ std::vector<TurnRestriction> Extractor::ParseOSMData(ScriptingEnvironment &scrip
     const RestrictionParser restriction_parser(
         scripting_environment.GetProfileProperties().use_turn_restrictions,
         config.parse_conditionals,
         restrictions);
 
-    // create a vector of iterators into the buffer
-    for (std::vector<osmium::memory::Buffer::const_iterator> osm_elements;
-         const osmium::memory::Buffer buffer = reader.read();
-         osm_elements.clear())
+    std::mutex process_mutex;
+
+    constexpr std::size_t BATCH_SIZE = 32;
+    std::vector<osmium::memory::Buffer> batch(BATCH_SIZE);
+    while (!reader.eof())
     {
-        for (auto iter = std::begin(buffer), end = std::end(buffer); iter != end; ++iter)
+        std::size_t batch_idx = 0;
+        while (batch_idx < batch.size() && (batch[batch_idx] = reader.read()))
         {
-            osm_elements.push_back(iter);
+            batch_idx++;
         }
 
-        // clear resulting vectors
-        resulting_nodes.clear();
-        resulting_ways.clear();
-        resulting_restrictions.clear();
-
-        scripting_environment.ProcessElements(osm_elements,
-                                              restriction_parser,
-                                              resulting_nodes,
-                                              resulting_ways,
-                                              resulting_restrictions);
-
-        number_of_nodes += resulting_nodes.size();
-        // put parsed objects thru extractor callbacks
-        for (const auto &result : resulting_nodes)
-        {
-            extractor_callbacks->ProcessNode(
-                static_cast<const osmium::Node &>(*(osm_elements[result.first])), result.second);
-        }
-        number_of_ways += resulting_ways.size();
-        for (const auto &result : resulting_ways)
-        {
-            extractor_callbacks->ProcessWay(
-                static_cast<const osmium::Way &>(*(osm_elements[result.first])), result.second);
-        }
-        number_of_relations += resulting_restrictions.size();
-        for (const auto &result : resulting_restrictions)
-        {
-            extractor_callbacks->ProcessRestriction(result);
-        }
+        tbb::parallel_for_each(
+            batch.begin(), batch.begin() + batch_idx, [&](const osmium::memory::Buffer &buffer) {
+                std::vector<std::pair<const osmium::Node &, ExtractionNode>> resulting_nodes;
+                std::vector<std::pair<const osmium::Way &, ExtractionWay>> resulting_ways;
+                std::vector<boost::optional<InputRestrictionContainer>> resulting_restrictions;
+
+                scripting_environment.ProcessElements(buffer,
+                                                      restriction_parser,
+                                                      resulting_nodes,
+                                                      resulting_ways,
+                                                      resulting_restrictions);
+
+                // serial insertion
```

Comment:
I think better is to decouple `ProcesElements` and `Process{Node,Way,Restriction}` like in [`flatten2d` example](https://www.threadingbuildingblocks.org/docs/help/reference/thread_local_storage/flattened2d_cls.html)

The idea is to use `enumerable_thread_specific` vectors for `resulting_nodes`, `resulting_ways`, `resulting_restrictions`  and finish parallel processing here.
And below use `parallel_do` with 3 jobs over flattened vectors with `ProcessNode`, `ProcessWay`, `ProcessRestrictions` calls. It should work without explicit locks, but `ProcessNode`, `ProcessWay`, `ProcessRestrictions` should not use any shared mutable state.